### PR TITLE
NCF Implementation in Tensorflow

### PR DESCRIPTION
**Pull request by AaronHeee**
### Description
Add NCF implementation in Tensorflow.

### Motivation and Context
Add neural-based recommender system for users.

### Releated Issues
[#363](https://github.com/Microsoft/Recommenders/issues/363) and [#339](https://github.com/Microsoft/Recommenders/issues/339)

### Checklist:

- [x] My code follows the code style of this project, as detailed in our [contribution guidelines](../CONTRIBUTING).
- [x] I have added tests.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by miguelgfierro** [comment]
This is awesome @AaronHeee, great contribution. The test might be failing due to the direct PR to master branch. The way we work in the team is to do a PR to staging, which is our development branch and then, after this branch is ready, we merge from staging to master and do a release. Do you mind to change the PR to staging?

File: reco_utils/recommender/ncf/__init__.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import datetime
+
```
[the remark this answers is not in the document]
Comment (AaronHeee) [reply reco_utils/recommender/ncf/__init__.py]: Thanks for your comments. 
I implement NCF based on the code of SAR, and don't notice that code in __init__.py. 
Sorry for that, I don't use data here, and will delete them soon.

File: reco_utils/recommender/ncf/dataset.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+        # sample n_neg negative samples for training
+        train_ratings = pd.merge(self.train, self.interact_status[[self.col_user, self.col_item + '_negative']],
+                                 on=self.col_user)
+
+        try:
+            train_ratings[self.col_item + '_negative'] = train_ratings[self.col_item + '_negative'].apply(
+                lambda x: random.sample(x, self.n_neg))
+        except:
+            raise ValueError
+            min_num = min(map(len, list(train_ratings[self.col_item + '_negative'])))
```
[the remark this answers is not in the document]
Comment (AaronHeee) [reply reco_utils/recommender/ncf/dataset.py]: This code is added in case the number of negative sampling `n_neg` is larger than the total number of negative items. If so, we will sample the negative items with smaller `n_neg` as the same as the smallest size of negative items.
Line 236 `raise ValueError` should be deleted, so the case is as the same as line 193. Thanks for your question, I will do it soon.

File: reco_utils/recommender/ncf/ncf_singlenode.py [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+        """
+
+        # get user and item mapping dict
+        self.user2id = data.user2id
+        self.item2id = data.item2id
+        self.id2user = data.id2user
+        self.id2item = data.id2item
+
+        # output the model type
+        log.info("Training model: %s" % self.model_type)
```
[the remark this answers is not in the document]
Comment (AaronHeee) [reply reco_utils/recommender/ncf/ncf_singlenode.py]: As the same as the first comment, I used log like SAR, but it's not enabled. I think that can be deleted if our code is only used in notebook.

File: reco_utils/recommender/ncf/ncf_singlenode.py [hunk 4]
```diff
@@ -0,0 +1,351 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
[the remark this answers is not in the document]
Comment (AaronHeee) [reply reco_utils/recommender/ncf/ncf_singlenode.py]: Thanks for your explanation of the notation of `_singlenode`. Because this implementation in Tensorflow run on CPU or single GPU, so I use this name `_singlenode` from SAR. I think it makes sense if other versions of ncf are planned.

**Comment by AaronHeee** [comment]
@miguelgfierro Thanks so much for your great comments and code review.
- Thanks for your tips, I have changed the branch from master to staging.
- This latex code runs correctly on my local notebook, I will solve this problem soon, thanks.
- I will add SVG figure, explain negative sampling and re-calculate these metrics for pre-trained ncf and update this notebook in the next few days.

**Comment by yueguoguo** [comment]
@miguelgfierro The hit ratio metric is used in the original NCF paper. Metrics in the repo are also reported in the notebook. Discussed with Aaron that we keep both in case readers are interested in any of them.

Negative sampling is a good concept used in implicit feedback scenario. I’ve put a new data_transform notebook to explain some of this. We may want to cover more under various algorithmic context.

File: reco_utils/recommender/ncf/ncf_singlenode.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (miguelgfierro) [reply reco_utils/recommender/ncf/ncf_singlenode.py]: makes sense

File: reco_utils/recommender/ncf/dataset.py [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+        self._init_test_data()
+
+    def _data_processing(self, train, test, implicit=True):
+        """ process the dataset to reindex userID and itemID, also set rating as implicit feedback
+
+        Args:
+            train (pandas.DataFrame): training data with at least columns (col_user, col_item, col_rating) 
+            test (pandas.DataFrame): test data with at least columns (col_user, col_item, col_rating)
+                    test can be None, if so, we only process the training data
+            implicit (bool): if true, set rating>0 to rating = 1
```
[the remark this answers is not in the document]
Comment (miguelgfierro) [reply reco_utils/recommender/ncf/dataset.py]: cool, thanks

File: reco_utils/recommender/ncf/ncf_singlenode.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (miguelgfierro) [reply reco_utils/recommender/ncf/ncf_singlenode.py]: cool, thanks

File: reco_utils/recommender/ncf/dataset.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (miguelgfierro) [reply reco_utils/recommender/ncf/dataset.py]: cool, thanks

File: reco_utils/recommender/ncf/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (miguelgfierro) [reply reco_utils/recommender/ncf/__init__.py]: cool, thanks

**Comment by miguelgfierro** [comment]
@AaronHeee @yueguoguo I think I know the reason why the tests are not being executed. They are executed when a person from MS do a PR from a branch in the repo to staging. Since Aaron is an external user and is doing a PR to staging, the tests are not executed for security reasons. We can do two things, we can manually check that the tests pass and merge or we can do like we did with Nicolas, do a PR to a branch created by us, and then Le or I do a PR to staging. 

This problem will be solved when we have the Azure devops pipeline in place, that Marcus is working on

**Comment by yueguoguo** [comment]
@miguelgfierro good catch! AFAIK @eisber is still working on the public branch build system. I have created a new branch named `ncf`. @AaronHeee you may need to change the target branch to that and then we can create a new one to merge into `staging` as Miguel suggested.

**Comment by WessZumino** [comment]
@AaronHeee and @yueguoguo thanks for the contribution! I also have a question concerning the hit ratio. In the notebook you write: 

*We truncated the ranked list at 10 for both metrics. As such, the HR intuitively measures whether the test item is present on the top-10 list* 

In my opinion, from reading this sentence, it is not clear how the hit ratio is different from the top_k. In both cases it would seem that a top-10 ranking is evaluated. Maybe it would be useful to better explain the difference between the two metrics amd how to compare them, as I believe they are not entirely unrelated. For example, the reported hit ration is 0.48 while precision@k is 0.17. How do I understand this difference?

File: reco_utils/recommender/ncf/dataset.py [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+                self.test,
+                test_interact_status[[self.col_user, self.col_item + '_negative']],
+                on=self.col_user,
+                how="left"
+            )
+
+            # sample n_neg_test negative samples for testing
+            try:
+                test_ratings[self.col_item + '_negative'] = test_ratings[self.col_item + '_negative'].apply(
+                    lambda x: random.sample(x, self.n_neg_test)
```
Comment (gramhagen) [inline reco_utils/recommender/ncf/dataset.py]: can you allow for setting random.seed() in constructor so that this sampling can be repeatable?

Comment (AaronHeee) [reply reco_utils/recommender/ncf/dataset.py]: Thanks for comments. I will set random seed. But there might be a small problem that negative sampling in every training epoch will be the same.  Do you have any suggestions for it?

File: tests/unit/test_ncf_dataset.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+            if is_pos:
+                assert i in positive_pool[u] 
+            else: 
+                assert i not in positive_pool[u] 
+
+    data.negative_sampling()
+
+    label_list = []
+
+    for shuffle in [True, False]:
```
Comment (gramhagen) [inline tests/unit/test_ncf_dataset.py]: what does this do?

File: reco_utils/recommender/ncf/dataset.py [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+        if shuffle:
+            np.random.shuffle(indices)
+        for i in range(len(indices) // batch_size):
+            begin_idx = i * batch_size
+            end_idx = (i + 1) * batch_size
+            batch_indices = indices[begin_idx: end_idx]
+            yield [
+                [self.id2user[x] for x in self.users[batch_indices]],
+                [self.id2item[x] for x in self.items[batch_indices]],
+                self.ratings[batch_indices],
```
Comment (gramhagen) [inline reco_utils/recommender/ncf/dataset.py]: it appears as though when this data is used it is converted back to ids? is that right? can we just keep it as ideas and pass it to TF?

File: reco_utils/recommender/ncf/ncf_singlenode.py [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+            pretrain=False,
+    ):
+        # number of users in dataset
+        self.n_users = n_users
+        # number of items in dataset
+        self.n_items = n_items
+        # model type
+        assert model_type.lower() in {"gmf", "mlp", "neumf"}
+
+        self.model_type = model_type
```
[the remark this answers is not in the document]
Comment (AaronHeee) [reply reco_utils/recommender/ncf/ncf_singlenode.py]: modified it, thanks!

File: reco_utils/recommender/ncf/dataset.py (hunk 6 above)
Comment (gramhagen) [reply reco_utils/recommender/ncf/dataset.py]: If you call
```random.seed(123)```
in the dataset constructor, then you should be able to reproduce the same results on the same dataset (provided the code doesn't change). So long as you aren't reinstantiating the object for each epoch, the random samples will be (repeatably) different each time you call random.sample()

**Comment by AaronHeee** [comment]
@miguelgfierro code updated. 

- [x] TF version, latex formula, SVG figure (yet it's broken in github jupyter notebook) and re-calculation of pre-trained ncf metrics.
- [ ] concept of negative sampling, @miguelgfierro will explain it in another notebook.

File: reco_utils/recommender/ncf/dataset.py (hunk 8 above)
Comment (AaronHeee) [reply reco_utils/recommender/ncf/dataset.py]: @gramhagen Yes, actually in the original version, we pass ids to TF directly. After discussing with @yueguoguo, we think function `train_loader()` could be called and used by our users in other situations , they expect the not re-indexed data. So we convert id --> user --> id to TF when using data loader. Yet that's slower.

**Comment by AaronHeee** [comment]
@WessZumino thanks so much for your question. I updated the notebook to explain the difference of HR@K in "leave-one-out" evaluation and Precision@K in generic evaluation.
We introduce the "leave-one-out" evaluation section to help users to understand and reproduce the original NCF paper. However, because the difference of data split, in our notebook, this isn't a **real** leave-one-out evaluation protocol. I explained it in new version notebook.

File: reco_utils/recommender/ncf/dataset.py (hunk 8 above)
Comment (gramhagen) [reply reco_utils/recommender/ncf/dataset.py]: I see, ok

**Comment by miguelgfierro** [comment]
@AaronHeee this is great. 

I saw a problem in the notebook, the image is not rendering at the moment. To fix it, you need to remove all this code:
```
from IPython.display import Image
from IPython.core.display import HTML 
Image(url= "img/NCF.svg", width=700, height=500)
```
and add the svg as an image in a markdown cell (not a code cell). A the bottom of this [notebook](https://github.com/miguelgfierro/sciblog_support/blob/master/Intro_to_Fraud_Detection/fraud_detection.ipynb) you can see an example of how I did it in the past, the line I used was: ``. I don't remember if it works with relative paths, instead of the url.

**Comment by AaronHeee** [comment]
@gramhagen @miguelgfierro, I updated code and fixed the beforementioned bugs. plz check it.
